### DISCUSSION TITLE
Download counts after Bintray -> GitHub move?

### DISCUSSION THREAD

**Discussion by bradcray** [Homebrew/discussions#1947] [Everyday usage]
Our project has historically maintained download statistics for our releases via homebrew and our other release mechanisms.  With [the move of hosted packages from Bintray to GitHub packages](https://github.com/Homebrew/discussions/discussions/691#discussioncomment-584299), we've lost our counts for Homebrew, not surprisingly.  Other than that interruption, we're generally very excited about this change.

With a quick look, we haven't discovered a way to use the GitHub API to get download counts for our package, and are considering scraping the HTML pages to get that value.  But before investing too much effort in this direction, we wanted to check and see whether others have come up with best practices for getting this information in the new world.

Thanks very much.

**Comment by SMillerDev** [comment]
There is an api at formulae.brew.sh. It'll give you install counts instead.

**Reply by bradcray** [reply]
Thanks for the super-fast response, @SMillerDev.  I'm pretty completely ignorant of formulae.brew.sh (and really, much of anything other than how to update our community-contributed formula from one release to the next).  I'm not seeing immediately where I should look—can you provide more precise coordinates?

**Reply by Rylan12** [reply]
HTML version: https://formulae.brew.sh/formula/wget

JSON version: https://formulae.brew.sh/api/formula/wget.json

Both have an analytics section that will give you install counts

**Reply by bradcray** [reply]
Thanks @Rylan12 for those pointers, which help.  We've traditionally displayed our download counts using [a line graph plotting accumulated downloads over time](https://chapel-lang.org/perf/download-stats/?startdate=2019/09/22&enddate=2020/04/03&graphs=chapel1200), which would require either wanting to compute how many installs were made each day (where we'd sum them up and zero the sum on each new release on our end), or how many installs were performed for a given formula / release (e.g., "version 1.xy was installed n times total").

With these rolling 30/90/365-day windows, my sense is that we'd be unable to get these figures since it would be impossible to know how many installs were made "yesterday" say, without also knowing how many installs were falling off the other end of the window.  Does that seem right?

If so, that makes me wonder whether there is any way to make a similar query for "the last 24 hours"?  Or better (in case our scripts were offline for a day) to get "total number of downloads of the formula for version 1.xy"?

Thanks again.

**Reply by Rylan12** [reply]
Hmm, I see. I know that when I go to Homebrew's GitHub packages explorer I can see approximate download counts (e.g. "8.32M"). I'm pretty sure those are publicly available.

For example, for [`chapel`](https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fchapel) (which I think is the formula you're referencing), I can see that there have been 648 total downloads, and also a breakdown of the number of downloads per-version (showing 75 for 1.24.1 and 20 for 1.24.0). Of course, 75 + 20 is not equal to 648 so I'm not sure what to make of that. You could probably scrape the webpage for these values without too much trouble if nothing else works.

I don't know if that's accessible via the GitHub API, though. It looks like you _can_ [get a package for an organization](https://docs.github.com/en/rest/reference/packages#get-a-package-for-an-organization) but it doesn't look like this shows the download count anywhere. I'm also not sure if you need to have elevated access to access this data.

@MikeMcQuaid or @sjackman might know more about whether this information is in the API somewhere and how to access it (and why the download counts don't seem to add up)

**Reply by bradcray** [reply]
Thanks @Rylan12.  Yes, Chapel is what I'm interested in, but was avoiding naming it so that whatever answers occurred here would be generally useful to others with the same question.

I was also confused by the 75 + 20 != 648 (where 648 is closer to the order of magnitude I'd expect since the GitHub conversion had taken place).  Another thing that's odd about that page is the mention that Chapel 1.24.1 was published 27 days ago when it was actually released on April 15th.

Ah, but checking its GitHub formula, I'm seeing there was an update to the formula [27 days ago](https://github.com/Homebrew/homebrew-core/commits/44d9fd8c793f664f70dada83a47502f43429f33d/Formula/chapel.rb).  So maybe dropping a new version of the package onto GitHub reset the 1.24.1 counts (?).

And running with that theory, I'm noting that if I click on the "All Versions" -> "8 untagged" link, I find the additional download counts that would help account for the 648:  https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fchapel/versions?filters%5Bversion_type%5D=untagged.  So that makes me think that perhaps there were 550 downloads of our original 1.24.1 formula, and then an additional 75 downloads of the revised formula from 27 days ago.

But, unfortunately, all of this makes me think that scraping the GitHub HTML to find the number of downloads of a given release won't be as straightforward an approach as I was hoping.  With the information I currently have, it seems like our best bet would be to count the total downloads daily, keep a running tally, and then when we move to version 1.25.0 (say), subtract the total downloads that we'd seen prior to the release from all future download counts as a rough approximation of how many 1.25.0 downloads there had been (which might be overcounting slightly if downloads of 1.24.x were still occurring).

So I feel like we're still hoping for a more straightforward / robust way to query "How many times have [all versions of] homebrew formula 1.xy been installed?"

**Reply by Rylan12** [reply]
Oh yeah, good catch. I bet you're right about the re-setting of the count. And agreed, that sounds like a reasonable way to do it if there's no better way. I'd also like to hear from Mike and Shaun (let's give them at least until Monday or Tuesday) because they know much more about GitHub packages than I do.

Another thing I'd add is that I think our use GitHub packages are still pretty new and so I wouldn't be surprised if such a thing is added in the future.

**Reply by SMillerDev** [reply]
You can always write a feature request for such an api to the github Roadmap repo.

**Reply by bradcray** [reply]
Hi @SMillerDev — Getting back to this, I was planning on filling such a feature request, but misunderstood that you were suggesting I post the request on GitHub's roadmap repo rather than homebrew's.  A challenge to doing this on the GitHub end would seem to be that a given tag that identifies a homebrew version (say `1.24.1`) can map to multiple packages on GitHub if minor updates are made to the formula during its lifetime, [like this one](https://github.com/Homebrew/homebrew-core/commit/44d9fd8c793f664f70dada83a47502f43429f33d#diff-0c9742fd008239155d7d6c8acb7ba65f603920010e388b6fe5c80e32e631d91b).  So somehow, GitHub would need to know "any of this set of packages represent this same logical version" which seems unlikely to be something they'd invest time into (so I'm hesitant to ask).

For others wrestling with this who might find their way to this issue, since my last note, I discovered that when viewing the page source for GitHub's top-level package pages ([like this one for wget](https://github.com/Homebrew/homebrew-core/pkgs/container/core%2Fwget), the raw number is available as a title string in the HTML:

```
<h3 title="1731526">1.73M</h3>
```

So a technique for calculating daily downloads would be to scrape that number daily.  And then each time a release is made, to make note of the previous day's number and subtract that value off of any subsequent counts.  (This is only approximately correct, of course, since if someone were to manually download older packages through GitHub, it would still get counted towards the current release's download counts.  But working under the assumption that hopefully few people are manually downloading old homebrew packages, that might be a reasonable simplifying assumption...)

Meanwhile, I was curious whether anyone knows how the analytics counts on [homebrew's formula pages](https://formulae.brew.sh/formula/wget) are computed?  I ask in part to try and understand the large disparity between the counts there and the ones on GitHub's pages; in part wondering whether there'd be any way to report them without a rolling window in order to use an approach similar to the one in the previous paragraph, but using data provided by homebrew rather than GitHub.

**Reply by SMillerDev** [reply]
> Meanwhile, I was curious whether anyone knows how the analytics counts on homebrew's formula pages are computed?

They're counting all install events that our Google Analytics records, but people might have this disabled.
